### ISSUE TITLE
Design an API for including preprocessing in `Experiment`

### ISSUE THREAD

**Issue by sharkovsky** [fedbiomed/fedbiomed#800]
As a developer, I want to introduce the pre-processing functionality in `Experiment`.
The first steps are:
- identify requirements
- design an API 
- implementation strategy

The deliverable of this task is a document including a description of the functionality, the requirements, and an API design. 
As a nice to have, we should provide a draft implementation strategy.

**Comment by sharkovsky** [comment]
## Summary of discussion with @ibalelli on 2023-06-14

### The pre-processing functionality includes (for now)
- standardization and normalization
- imputation

### Intended use cases
1. Researcher wants to run a federated experiment with a specific ML task, and they want to use preprocessing to manage the input data to their task
2. Clinician wants to run a preprocessing pipeline on their local dataset (mostly interested in imputation) and be able to save their results locally
3. Researcher wants to train a specific preprocessing method, e.g. they want to train an autoencoder, with the goal of using it for use cases 1 and 2 later on

### Supported preprocessing methods
As a POC, we would like to be able to support:
| method | required |
| --- | --- |
| Fed mean imputation | yes |
| Fed-MIWAE imputation | yes |
| Fed mean/stddev normalization | yes |
| Santi's normalization | nice to have |
| PPCA | nice to have |
| Aude's GAN | nice to have |

### Other notes
- Some preprocessing methods may require training themselves. In addition, almost all of them will require at least a first round of communication (e.g. to compute federated mean). In simple cases, this may be handled automatically by the experiment class, but in complex cases (e.g. training an autoencoder) we assume that it will be done separately as explained in use case 3.
- For use case 1, data may be preprocessed on-the-fly as part of the dataloader. However, we must be able to guarantee determinism, i.e. one missing data point should always be imputed with the same value (not trivial with MIWAE)
- Supporting PPCA requires significant work for integrating it into Fed-BioMed (it uses EM and is not implemented in torch).
- Aude has been working on a GAN method as a benchmark to compare for imputation. Supporting GANs is potentially complicated in Fed-BioMed, although she has already made a poc about it. 
- For the **paper**, we would like to provide a reproducibility bundle with the necessary code to reproduce Fed-MIWAE
- Use case 2 corresponds to the situation where a clinician wants to run the imputation locally, possibly multiple times (without the determinism requirement of use case 1), and then look at the distribution of their data. It would be nice to somehow allow this via the gui. 
- A draft first list of parameters for the preprocessing includes: which data to run it on, which method to use,  method hyperparameters

**Comment by sharkovsky** [comment]
# Summary of discussion on 2023-06-20

Participants: @marcolorenzi, @ibalelli, @sssilvar , Aude

### The pre-processing functionality should include
- query to obtain more fine-grained information than what we provide right now, e.g. descriptive statistics, percent missing data, etc. (think of `pandas.describe`)
- potentially, some basic visualization like histograms
- the two points above may, in the future, be associated with a privacy module (e.g. adding laplace noise to histogram bins) 
- imputation
- other data-modifying operations: standardization, normalization, bias correction, etc..

We also discussed the possibility to include, in the future, some post-processing to allow researchers to perform "model debugging". However this may be too complicated for now (but might revisit later).

### Intended use cases
1. Researcher wants to obtain descriptive information about the datasets, e.g. aggregate statistics, distributions, etc..
2. Researcher wants to run a federated experiment with a specific ML task, and they want to use preprocessing to manage the input data to their task
3. Clinician wants to run a preprocessing pipeline on their local dataset (mostly interested in imputation) and be able to save their results locally
4. Researcher wants to train a specific preprocessing method, e.g. they want to train an autoencoder, with the goal of using it for use cases 1 and 2 later on

### Supported methods
As a POC, we would like to be able to support:
| method | required |
| --- | --- |
| federated descriptive statistics | yes |
| federated histogram | yes |
| Fed mean imputation | yes |
| Fed-MIWAE imputation | yes |
| Fed mean/stddev normalization | yes |
| Santi's normalization | nice to have |
| PPCA | nice to have |
| Aude's GAN | nice to have |

### Other notes
- The functions to describe the datasets can (or even should?) be built-in to Fed-BioMed. This may provide some protection against the obvious threats of revealing too much information about the data (unless we provide a private way of accessing the dataset)
- the preprocessing functions may be built-in or provided by the researcher
- We all recognize the issue of avoiding saving data on the node. The option of imputing data on the fly is acceptable, but may be computationally expensive. We also retain the possibility that nodes can execute the imputation offline, through a separate mechanism that is not shared with the researcher. In this case, it should be ok to save data on the node filesystem. Later, they may share the dataset with imputation as they would do normally with other datasets.
- the point above seems to imply that we would need to keep some record (e.g. a registry) of imputation models that can be executed by the node 
- continuing the situation from the point above, a useful feature for the node to have is being able to distinguish whether a certain preprocessing model has been trained or not yet on their dataset, to know whether they can apply imputation or not.
- the kinds of descriptive information that is provided may be dependent on the type of dataset. In particular, images and tabular data may output different things to the same request of "describing" the node's distribution
- we may want to add the functionality to perform preprocessing only on datasets that satisfy a certain condition

### Additional points that were not discussed
A couple of things that I just thought about while writing these notes.

- maybe we should implement a minimum sample size threshold for descriptive information, e.g. nothing is returned if the dataset has less than 20 samples
- yet another option for running imputation could be a request mechanism similar to what we have for training plan approval. The researcher submits a request for imputation, and the node may perform the imputation offline and save the new dataset. This would save us the cost of keeping a registry of imputation models that may be executed by the node, since the researcher would insert a reference to a specific model directly in their request

**Comment by mvesin** [comment]
Hi @ybouilla 
I think we mentioned and agreeed that disk usage is one of the reason for not duplicating datasets, during this morning's discussion.
Best regards,

**Comment by sharkovsky** [comment]
The documents in this [folder](https://mybox.inria.fr/smart-link/8efa3cae-5671-45fa-8337-ee25026f3bd3/) (internal inria link) are an updated collection of our discussion.
They are currently being reviewed by @ibalelli and other interested parties.
